keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,453 @@
+       On a scale down followed by a scale up, will wait until the old PVC for the removed Pod is deleted and ensure 
+       that the PVC used is a freshly created one.
+   * `DeleteOnStatefulSetDeletion` - PVCs corresponding to the StatefulSet are deleted when StatefulSet
+       themselves get deleted. When Pods are deleted as part of a scale down, PVCs are not
+       deleted. Thus there may be PVCs owned by the StatefulSet that are not attached to a Pod.
+2. Add `patch` to the statefulset controller rbac cluster role for `persistentvolumeclaims`.
+
+### User Stories
+
+#### Story 1
+The user is running a StatefulSet as part of an application with a finite lifetime. During
+the application's existence the StatefulSet maintains per-pod state, even across scale-up
+and scale-down. In order to maximize performance, volumes are retained during scale-down
+so that scale-up can leverage the existing volumes. When the application is finished, the
+volumes created by the StatefulSet are no longer needed and can be automatically
+reclaimed.
+
+The user would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnStatefulSetDelete`
+which would ensure that the PVCs created automatically during the StatefulSet activation
+is deleted once the StatefulSet is deleted.
+
+#### Story 2
+User is very cost conscious and can sustain slower scale-up speeds, even after a
+scale-down. The user does not want to pay for volumes that are not in use, and so wants
+them to be reclaimed as soon as possible, including during scale-down. On scale-up a new
+volume will be provisioned and the new pod will have to re-intitialize. However, for
+short-lived interruptions like a rolling update or after pod rescheduling, the data on
+volumes is persisted. This is a key property that ephemeral storage, like emptyDir, cannot
+provide.
+
+User would set the `PersistentVolumeClaimDeletePolicy` as `DeleteOnScaledown` ensuring
+PVCs are deleted when corresponding Pods are deleted. New Pods created during scale-up
+followed by a scale-down will wait for freshly created PVCs.
+
+### Notes/Constraints/Caveats (optional)
+
+This feature applies to PVCs which are dynamically provisioned from the volumeClaimTemplate of 
+a StatefulSet. Any PVC and PV provisioned from this mechanism will function with this feature.
+
+### Risks and Mitigations
+
+Currently the PVCs created by StatefulSet are not deleted automatically. Using the
+`DeleteOnScaledown` or `DeleteOnStatefulSetDeletion` would delete the PVCs
+automatically. Since this involves persistent data being deleted, users should take
+appropriate care using this feature. Having the `Retain` behaviour as default will ensure
+that the PVCs remain intact by default and only a conscious choice made by user will
+involve any persistent data being deleted.
+
+PVCs associated with the StatefulSet will be more durable than ephemeral volumes would be,
```

Comment:
There was a request to clarify why this feature was needed, because it seems to duplicate ephemeral volumes (generally, including emptyDir). I've tried to make the context more clear.